sled-agent/src/swap_device.rs

```diff
@@ -0,0 +1,504 @@
+            }
+        }
+    }
+
+    // The argument for SC_LIST (struct swaptbl) requires an embedded array in
+    // the struct, with swt_n entries, each of which requires a pointer to store
+    // the path to the device.
+    //
+    // Ideally, we would want to query the number of swap devices on the system
+    // via SC_GETNSWP, allocate enough memory for each device entry, then pass
+    // in pointers to memory to the list command. Unfortunately, creating a
+    // generically large array embedded in a struct that can be passed to C is a
+    // bit of a challenge in safe Rust. So instead, we just pick a reasonable
+    // max number of devices to list.
+    //
+    // We pick a max of 3 devices, somewhat arbitrarily. We only ever expect to
+    // see 0 or 1 swap device(s); if there are more, that is a bug. In the case
+    // that we see more than 1 swap device, we log a warning, and eventually, we
+    // should send an ereport.
+    const N_SWAPENTS: usize = 3;
+
+    // Wrapper around swapctl(2) call. All commands except SC_GETNSWP require an
+    // argument, hence `data` being an optional parameter.
+    unsafe fn swapctl_cmd<T>(
+        cmd: i32,
+        data: Option<std::ptr::NonNull<T>>,
+    ) -> std::io::Result<u32> {
+        assert!(
+            cmd >= SC_ADD && cmd <= SC_GETNSWP,
+            "invalid swapctl cmd: {cmd}"
+        );
+
+        let ptr = match data {
+            Some(v) => v.as_ptr() as *mut libc::c_void,
+            None => std::ptr::null_mut(),
+        };
+
+        let res = swapctl(cmd, ptr);
+        if res == -1 {
+            return Err(std::io::Error::last_os_error());
+        }
+
+        Ok(res as u32)
+    }
+
+    /// List swap devices on the system.
+    pub(crate) fn list_swap_devices() -> Result<Vec<SwapDevice>, SwapDeviceError>
+    {
+        // Statically create the array of swapents for SC_LIST; see comment on
+        // `N_SWAPENTS` for details as to why we do this statically.
+        //
+        // Each swapent requires a char * pointer in our control for the
+        // `ste_path` field,, which the kernel will fill in with a path if there
+        // is a swap device for that entry. Because these pointers are mutated
+        // by the kernel, we mark them as mutable. (Note that the compiler will
+        // happily accept these definitions as non-mutable, since it can't know
+        // what happens to the pointers on the C side, but not marking them as
+        // mutable when they may be in fact be mutated is undefined behavior).
+        //
+        // Per limits.h(3HEAD), PATH_MAX is the max number of bytes in a path
+        // name, including the null terminating character, so these buffers
+        // have sufficient space for paths on the system.
+        const MAXPATHLEN: usize = libc::PATH_MAX as usize;
+        assert_eq!(N_SWAPENTS, 3);
+        let mut p1 = [0i8; MAXPATHLEN];
+        let mut p2 = [0i8; MAXPATHLEN];
+        let mut p3 = [0i8; MAXPATHLEN];
```

Comment:
I'm okay with the constant here, but couldn't we still initialize these relative to the `N_SWAPENTS` constant?

What do you think about something like:

```rust
let mut pathptrs = [[0i8; MAXPATHLEN]; N_SWAPENTS];
let entries: [swapent; N_SWAPENTS] = pathptrs.iter_mut()
  .map(|p| swapent { ste_path: p as *mut libc::c_char, ..Default::default() })
  .collect::<Vec<_>>()
  .try_into()
  .expect("Mapping array to vec of same size");
```

Reply:
I think we could, but I don't expect this to change any time soon, and to me the way I wrote it makes things a bit clearer with regards to how the underlying memory is used here. Mind if I leave as is?